doc/design.md

```diff
@@ -1,51 +1,203 @@
 # Design
 
-## Project Structure Definition
-`dsn~project-structure-definition~1`
+## Validators
 
-The project structure is defined by a file structure in the resources folder of this plugin.
-There are two folders: `require_exist` and `require_exact`. 
-For files in the `require_exist` folder, this plugin only checks if they exist.
-If they don't exist it creates the file with the content from the file in `require_exist` as a template.
+`dsn~validators~1`
 
-For the files in the `require_exact` this plugin also validates that the content is exactly the same.
+As an interface for the different validations we created the `Validator` interface. It only contains the `validate` method:
 
-Rationale:
-This way of defining the project structure makes it very convenient to add new content.
-Now knowledge about the internal implementation of this plugin is required.
-That allows users to quickly add or modify the template.
+```java
+public List<ValidationFinding> validate();
+```
+
+The `ValidationFinding` consists of:
+
+* A message
+* Optionally: A fix for the finding (function closure that changes the project so that the finding will not happen again)
 
-It is not a problem that both folders overlap. In case a file is defined in both directories, 
-this plugin validates it as if it would only exist in `require_exact`. 
+We decided for this interface since it allows us to:
+
+* Report errors that can't be fixed automatically
+* Report error and directly provide a fix for it
+
+A problem with this approach is that some validation-fixes change a resource (typically files). If a validator creates multiple validation findings for the same resource, each fix needs to read and write the file. Consider for example the plugin validator for the pom file. If the pom file validator detects multiple missing plugins, each of the fixes would need to write the file. Since the fixes are executed by the surrounding code and not in a specific order. To solve this, we used the trick to create a compound finding. So basically one finding with one error message that combines the one of all it's sub findings. Its fix is a method that runs the fixes of all sub-findings and then writes the pom file.
```

Comment:
```suggestion
A problem with this approach is that some validation fixes change a resource such as files. If a validator creates multiple validation findings for the same resource, each fix needs to read and write the file. 

Consider, for example, the plugin validator for the pom file. If the pom file validator detects multiple missing plugins, each of these fixes would need to write the file since the fixes are executed by the surrounding code and not in a specific order. To solve this, we used the trick to create a compound finding. So basically one finding with one error message that combines one of all its sub findings. Its fix is a method that runs the fixes of all sub-findings and then writes the pom file.
```